In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Master Path

path = r'/Users/jagpreetbatth/Desktop/Career Foundry/Python/11-2021 Instacart Basket Analysis'

In [3]:
# Verify Path

path

'/Users/jagpreetbatth/Desktop/Career Foundry/Python/11-2021 Instacart Basket Analysis'

In [4]:
# Import orders_products_merged

df_ords_prods_merged = pd.read_pickle(os.path.join(path, 'Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [5]:
# Verify Correct Import

df_ords_prods_merged.shape

(32641268, 14)

### Implementing a user-defined function in the Instacart project

In [6]:
# Create subset of df_ords_prods_merged (due to large data load)

df = df_ords_prods_merged[:1000000]

In [7]:
# Verify Correct Subset Defined

df.shape

(1000000, 14)

In [8]:
df

order_id  user_id  order_number  orders_day_of_week  \
0        2398795      1.0           2.0                 3.0   
1        2398795      1.0           2.0                 3.0   
2        2398795      1.0           2.0                 3.0   
3        2398795      1.0           2.0                 3.0   
4        2398795      1.0           2.0                 3.0   
...          ...      ...           ...                 ...   
999995   1270896   6833.0           5.0                 2.0   
999996   1270896   6833.0           5.0                 2.0   
999997   1270896   6833.0           5.0                 2.0   
999998   1270896   6833.0           5.0                 2.0   
999999   1270896   6833.0           5.0                 2.0   

        order_hour_of_day  days_since_prior_order  product_id  \
0                     7.0                    15.0       196.0   
1                     7.0                    15.0     10258.0   
2                     7.0                    15.0     12427.0   
3                     7.0                    15.0     13176.0   
4                     7.0                    15.0     26088.0   
...                   ...                     ...         ...   
999995               12.0                    20.0     19836.0   
999996               12.0                    20.0      6155.0   
999997               12.0                    20.0     17993.0   
999998               12.0                    20.0     17896.0   
999999               12.0                    20.0     15134.0   

        add_to_cart_order  reordered                            product_name  \
0                     1.0        1.0                                    Soda   
1                     2.0        0.0                              Pistachios   
2                     3.0        1.0                     Original Beef Jerky   
3                     4.0        0.0                  Bag of Organic Bananas   
4                     5.0        1.0              Aged White Cheddar Popcorn   
...                   ...        ...                                     ...   
999995                1.0        0.0        Original Breakfast Sausage Links   
999996                2.0        0.0                  Original String Cheese   
999997                3.0        0.0  Shredded Colby & Monterey Jack Cheeses   
999998                4.0        0.0    Sugar Free Lemon Lime Gelatin Snacks   
999999                5.0        1.0                      Strawberry Gelatin   

        aisle_id  department_id  prices _merge  
0           77.0            7.0     9.0   both  
1          117.0           19.0     3.0   both  
2           23.0           19.0     4.4   both  
3           24.0            4.0    10.3   both  
4           23.0           19.0     4.7   both  
...          ...            ...     ...    ...  
999995     106.0           12.0    15.9   both  
999996      21.0           16.0     5.5   both  
999997      21.0           16.0    10.0   both  
999998      71.0           16.0    14.5   both  
999999      71.0           16.0     7.6   both  

[1000000 rows x 14 columns]

In [9]:
# Define Conditions based on Provided Price Ranges using if-else constructs

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [10]:
# Application of Function

df['price_range'] = df.apply(price_label, axis=1)

/var/folders/19/g__32h_54bjf6q9mxl_6vfjr0000gn/T/ipykernel_45562/380887843.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [11]:
# Check values of newly created price label column

df['price_range'].value_counts(dropna = False)

Mid-range product    667457
Low-range product    312587
High range            12209
Not enough data        7747
Name: price_range, dtype: int64

In [12]:
# Most expensive product within the subset

df['prices'].max()

99999.0

In [13]:
# Define Conditions based on Provided Price Ranges using loc() Function

df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

/opt/anaconda3/lib/python3.9/site-packages/pandas/core/indexing.py:1684: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[key] = infer_fill_value(value)
/opt/anaconda3/lib/python3.9/site-packages/pandas/core/indexing.py:1817: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value, pi)


In [14]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [15]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [16]:
# Check values of newly created price label column

df['price_range_loc'].value_counts(dropna = False)

Mid-range product     667457
Low-range product     312587
High-range product     12209
NaN                     7747
Name: price_range_loc, dtype: int64

In [17]:
# Use loc() function on entire dataframe to avoid memory overload faliure

df_ords_prods_merged.loc[df_ords_prods_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [18]:
df_ords_prods_merged.loc[(df_ords_prods_merged['prices'] <= 15) & (df_ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [19]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [20]:
# Check values of newly created price label column for entire dataframe

df_ords_prods_merged['price_range_loc'].value_counts(dropna = False)

Mid-range product     21860860
Low-range product     10126321
High-range product      417678
NaN                     236409
Name: price_range_loc, dtype: int64

In [21]:
# Which day most orders take place - frequency of orders_day_of_the_week

df_ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

0.0    5827441
1.0    5339862
6.0    4224489
5.0    3983274
2.0    3976654
3.0    3628534
4.0    3582912
NaN    2078102
Name: orders_day_of_week, dtype: int64

In [22]:
# Create a for-loop to create a busiest day column

result = []

for value in df_ords_prods_merged["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [23]:
# Check Result

result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regular

In [24]:
# Merge result with ords_prods_merge

df_ords_prods_merged['busiest_day'] = result

In [25]:
# Check values of newly created busiest day column

df_ords_prods_merged['busiest_day'].value_counts(dropna = False)

Regularly busy    23230915
Busiest day        5827441
Least busy         3582912
Name: busiest_day, dtype: int64

In [26]:
# Verify busiest day column creation

df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2398795      1.0           2.0                 3.0                7.0   
1   2398795      1.0           2.0                 3.0                7.0   
2   2398795      1.0           2.0                 3.0                7.0   
3   2398795      1.0           2.0                 3.0                7.0   
4   2398795      1.0           2.0                 3.0                7.0   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    15.0       196.0                1.0        1.0   
1                    15.0     10258.0                2.0        0.0   
2                    15.0     12427.0                3.0        1.0   
3                    15.0     13176.0                4.0        0.0   
4                    15.0     26088.0                5.0        1.0   

                 product_name  aisle_id  department_id  prices _merge  \
0                        Soda      77.0            7.0     9.0   both   
1                  Pistachios     117.0           19.0     3.0   both   
2         Original Beef Jerky      23.0           19.0     4.4   both   
3      Bag of Organic Bananas      24.0            4.0    10.3   both   
4  Aged White Cheddar Popcorn      23.0           19.0     4.7   both   

     price_range_loc     busiest_day  
0  Mid-range product  Regularly busy  
1  Low-range product  Regularly busy  
2  Low-range product  Regularly busy  
3  Mid-range product  Regularly busy  
4  Low-range product  Regularly busy

In [27]:
# Create for-loop specifying busiest day(s)

result_two = []

for value in df_ords_prods_merged["orders_day_of_week"]:
  if value == 0 or value == 1:
    result_two.append("Busiest days")
  elif value == 3 or value == 4:
    result_two.append("Least busy days")
  else:
    result_two.append("Regularly busy")

In [29]:
# Check Result_two

result_two

['Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy da

In [30]:
# Create new column for busiest day(s)

df_ords_prods_merged['busiest_days'] = result_two

In [31]:
# Verify busiest day(s) column creation

df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2398795      1.0           2.0                 3.0                7.0   
1   2398795      1.0           2.0                 3.0                7.0   
2   2398795      1.0           2.0                 3.0                7.0   
3   2398795      1.0           2.0                 3.0                7.0   
4   2398795      1.0           2.0                 3.0                7.0   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    15.0       196.0                1.0        1.0   
1                    15.0     10258.0                2.0        0.0   
2                    15.0     12427.0                3.0        1.0   
3                    15.0     13176.0                4.0        0.0   
4                    15.0     26088.0                5.0        1.0   

                 product_name  aisle_id  department_id  prices _merge  \
0                        Soda      77.0            7.0     9.0   both   
1                  Pistachios     117.0           19.0     3.0   both   
2         Original Beef Jerky      23.0           19.0     4.4   both   
3      Bag of Organic Bananas      24.0            4.0    10.3   both   
4  Aged White Cheddar Popcorn      23.0           19.0     4.7   both   

     price_range_loc     busiest_day     busiest_days  
0  Mid-range product  Regularly busy  Least busy days  
1  Low-range product  Regularly busy  Least busy days  
2  Low-range product  Regularly busy  Least busy days  
3  Mid-range product  Regularly busy  Least busy days  
4  Low-range product  Regularly busy  Least busy days

In [33]:
# Study Numbers for busiest and least busiest day(s) 0=SAT, 1=SUN, 2=MON, 3=TUE, 4=WED, 5=THU, 6=FRI

df_ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

0.0    5827441
1.0    5339862
6.0    4224489
5.0    3983274
2.0    3976654
3.0    3628534
4.0    3582912
NaN    2078102
Name: orders_day_of_week, dtype: int64

In [32]:
# Check values of newly created busiest day(s) column

df_ords_prods_merged['busiest_days'].value_counts(dropna = False)

Regularly busy     14262519
Busiest days       11167303
Least busy days     7211446
Name: busiest_days, dtype: int64

# Check for Accuracy of Results 
## The count above shows these columns are accurate. Count for the least busy day(s) (3,4) is 3,628,534 + 3,582,912 which equals to 7,211,446 and count for the most busy day(s) (0,1) is 5,827,441 + 5,339,862  which equals 11,167,303.  This proves SAT and SUN are the BUSIEST DAYS and TUES and WED are the LEAST BUSIEST DAY(S).

In [34]:
# Busiest Hours of the Day

df_ords_prods_merged['order_hour_of_day'].value_counts(dropna = False)

10.0    2612807
11.0    2583662
14.0    2537019
15.0    2507345
13.0    2506556
12.0    2464685
16.0    2383725
9.0     2328055
NaN     2078102
17.0    1960305
8.0     1634048
18.0    1534215
19.0    1179165
20.0     917028
7.0      850713
21.0     751626
22.0     596846
23.0     378604
6.0      276824
0.0      205013
1.0      108995
5.0       83402
2.0       64486
4.0       49794
3.0       48248
Name: order_hour_of_day, dtype: int64

In [35]:
# Create for-loop specifying busiest hours of the day

result_three = []

for value in df_ords_prods_merged["order_hour_of_day"]:
  if value == 10 or value == 11 or value == 14 or value == 15 or value == 13 or value == 12 or value == 16 or value == 9 == 17 or value == 8 or value == 18 or value == 19:
    result_three.append("Most orders")
  elif value == 20 or value == 7 or value == 21 or value == 22 or value == 23 or value == 6 or value == 0 or value == 1:
    result_three.append("Average orders")
  else:
    result_three.append("Fewest orders")

In [36]:
# Check Result_three

result_three

['Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 

In [37]:
# New Column for Busiest hour of the day

# Create new column for busiest period of day
df_ords_prods_merged['busiest_period_of_day'] = result_three

In [38]:
# Verify Column Creation 

df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2398795      1.0           2.0                 3.0                7.0   
1   2398795      1.0           2.0                 3.0                7.0   
2   2398795      1.0           2.0                 3.0                7.0   
3   2398795      1.0           2.0                 3.0                7.0   
4   2398795      1.0           2.0                 3.0                7.0   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    15.0       196.0                1.0        1.0   
1                    15.0     10258.0                2.0        0.0   
2                    15.0     12427.0                3.0        1.0   
3                    15.0     13176.0                4.0        0.0   
4                    15.0     26088.0                5.0        1.0   

                 product_name  aisle_id  department_id  prices _merge  \
0                        Soda      77.0            7.0     9.0   both   
1                  Pistachios     117.0           19.0     3.0   both   
2         Original Beef Jerky      23.0           19.0     4.4   both   
3      Bag of Organic Bananas      24.0            4.0    10.3   both   
4  Aged White Cheddar Popcorn      23.0           19.0     4.7   both   

     price_range_loc     busiest_day     busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy  Least busy days        Average orders  
1  Low-range product  Regularly busy  Least busy days        Average orders  
2  Low-range product  Regularly busy  Least busy days        Average orders  
3  Mid-range product  Regularly busy  Least busy days        Average orders  
4  Low-range product  Regularly busy  Least busy days        Average orders

In [39]:
# Frequency of new Column

df_ords_prods_merged['busiest_period_of_day'].value_counts(dropna = False)

Most orders       21943227
Fewest orders      6612392
Average orders     4085649
Name: busiest_period_of_day, dtype: int64

In [41]:
# Verify Correct Format and Dimension

df_ords_prods_merged.shape

(32641268, 18)

In [42]:
# Export dataframe as a pickle file into Prepared Data Folder

df_ords_prods_merged.to_pickle(os.path.join(path, 'Data','Prepared Data', 'ords_prods_added_columns.pkl'))